controller/lib/core/blower_fsm.h

```diff
@@ -159,19 +172,125 @@ class PressureControlFsm {
+// This mode enforce a minimum respiratory rate in case the patient does not
+// initiate enough breaths.
+// Similarly to pressure control, uses a square wave, but the duration of the
+// exhale leg can vary depending on the patient breath initiation, meaning the
+// breath can be finished (and the next one started) before the full breath
+// duration has passed.
+// In case the full breath duration passes, the fsm enforces a "mandatory
+// breath" to make sure the patient's RR does not go below the desired one.
+// Because the inhale is the first leg and only the exhale leg can vary in
+// duration, the first breath in this mode is always a "mandatory breath".
+class PressureAssistFsm {
+public:
+  explicit PressureAssistFsm(Time now, const VentParams &params)
+      : inspire_pressure_(cmH2O(static_cast<float>(params.pip_cm_h2o))),
+        expire_pressure_(cmH2O(static_cast<float>(params.peep_cm_h2o))),
+        start_time_(now), inspire_end_(start_time_ + InspireDuration(params)),
+        latest_expire_end_(inspire_end_ + MaxExpireDuration(params)) {}
+
+  void Update(Time now, const SensorReadings &readings) {
+    if (now >= inspire_end_) {
+      inspire_finished_ = true;
+    }
+    if (now > latest_expire_end_ || PatientInspiring(readings)) {
+      finished_ = true;
+    }
+  }
+
+  BlowerSystemState DesiredState() const {
+    if (inspire_finished_) {
+      return {.blower_enabled = true, expire_pressure_, ValveState::OPEN};
+    }
+    return {.blower_enabled = true, inspire_pressure_, ValveState::CLOSED};
+  }
+
+  bool Finished() const { return finished_; }
+
+private:
+  // Given t = secs_per_breath and r = I:E ratio, calculate inspiration duration
+  // and maximum expiration durations (I and E).
+  static Duration InspireDuration(const VentParams &params) {
+    float t =
+        60.f / static_cast<float>(params.breaths_per_min); // secs per breath
+    float r = params.inspiratory_expiratory_ratio;         // I:E
+    return seconds(t * r / (1 + r));
+  }
+  static Duration MaxExpireDuration(const VentParams &params) {
+    float t =
+        60.f / static_cast<float>(params.breaths_per_min); // secs per breath
+    float r = params.inspiratory_expiratory_ratio;         // I:E
+    return seconds(t / (1 + r));
+  }
+
+  bool PatientInspiring(const SensorReadings &readings) {
+    // TODO: change this when we get better understanding of inspiratory effort
+    // detection strategy, right now this is just set to trigger for any flow
+    // that is higher than twice the flowmeter's flow noise :
+    //   flow = flow_inhale (+/- noise) - flow_exhale (+/- noise).
+    // According to data from Ingmar test lung
+    // (https://drive.google.com/open?id=11_A8KEDdBa5l7rqUegfKsj71rug1l7Sx),
+    // this should work OK as a first approximation.
+    if (inspire_finished_ && readings.flow_ml_per_min >= 0 &&
+        inspiratory_effort_threshold_ >= cubic_m_per_sec(9)) {
```

Comment:
Please document where the cubic_m_per_sec(9) comes from. I hope the flow noise is not 9 m^3/sec?